genomic_address_service/classes/multi_level_clustering.py

```diff
@@ -1,61 +1,234 @@
 import numpy as np
 import scipy
+import skbio.tree
 
 class multi_level_clustering:
+    """
+    Perform hierarchical clustering with multiple threshold levels.
 
-    def __init__(self,dist_mat_file,thresholds,method):
-        self.labels, matrix = self.read_distance_matrix(dist_mat_file)
+    This class wraps around SciPy's hierarchical clustering tools to:
+    - Read a precomputed distance matrix.
+    - Generate a linkage matrix.
+    - Assign cluster memberships at multiple distance thresholds.
+    - Export the resulting tree in Newick format.
+
+    Attributes
+    ----------
+    cluster_memberships : dict[str, list[str]]
+        Mapping from each label to a list of cluster IDs across thresholds.
+    thresholds : list of float
+        Distance thresholds at which the dendrogram is cut to form clusters.
+    labels : list of str
+        Labels corresponding to the original observations (tips).
+    linkage : np.ndarray
+        The hierarchical clustering linkage matrix of shape (n-1, 4).
+    newick : str
+        The Newick-formatted string representation of the tree.
+
+    Notes
+    -----
+    - Cluster memberships are stored as strings, one per threshold.
+    - The linkage matrix is built using SciPy's `scipy.cluster.hierarchy.linkage`.
+    - Newick export relies on scikit-bio's `TreeNode`.
+    """
+    VALID_TREE_DISTANCES = ['patristic', 'cophenetic']
+
+    def __init__(self, dist_mat_file, thresholds, method, tree_distances='patristic'):
+        """
+        Initialize the clustering object.
+
+        Parameters
+        ----------
+        dist_mat_file : str
+            Path to a tab-delimited distance matrix file.
+        thresholds : list of float
+            Distance thresholds at which to assign clusters.
+        method : str
+            Linkage method to use (e.g., 'single', 'complete', 'average', 'ward').
+        tree_distances : string, optional (default=patristic)
+            Defines how distances in the tree (Newick file) correspond to distances in the original
+            distance matrix used to construct the tree.
+            If 'patristic', the patristic distance between leaves of the tree correspond to the distance matrix.
+            If 'cophenetic', the cophenetic distance between leaves (height in tree where leaves first
+            share a common ancestor) correspond to distances in the original distance matrix. For ultrametic trees
+            this corresponds to twice the patristic distance.
```

Comment:
I'd reverse the "correspond to" order here. For example:

If pastristic, the distances in the distance matrix correspond to the patristic distances between nodes in the tree

Also patristic distances can be measured between internal nodes as well, so I would change leaves to nodes here, especially since this isn't user-facing.

Reply:
Thanks so much. I've fixed up this text in f044599f9cc493afd9cb46ab2bc4a71894157944